test/unit/cursor/find_cursor.test.js

```diff
@@ -10,6 +10,115 @@ const { FindCursor } = require('../../../src/cursor/find_cursor');
 
 const test = {};
 describe('Find Cursor', function () {
+  describe('#next', function () {
+    afterEach(function () {
+      mock.cleanup();
+    });
+    beforeEach(function () {
+      return mock.createServer().then(mockServer => {
+        test.server = mockServer;
+      });
+    });
+
+    context('when there is a data bearing server', function () {
+      beforeEach(function () {
+        test.server.setMessageHandler(request => {
+          const doc = request.document;
+          if (isHello(doc)) {
+            request.reply(mock.HELLO);
+          } else if (doc.find) {
+            request.reply({
+              cursor: {
+                id: Long.fromNumber(1),
+                ns: 'test.test',
+                firstBatch: [{ _id: 1, name: 'test' }]
+              },
+              ok: 1
+            });
+          }
+        });
+      });
+
+      it('sets the session on the cursor', function (done) {
+        const topology = new Topology(test.server.hostAddress());
+        const cursor = new FindCursor(topology, MongoDBNamespace.fromString('test.test'), {}, {});
+        topology.connect(function () {
+          cursor.next(function () {
+            expect(cursor.session).to.exist;
+            topology.close(done);
+          });
+        });
+      });
+    });
+
+    context('when there is no data bearing server', function () {
+      beforeEach(function () {
+        test.server.setMessageHandler(request => {
+          const doc = request.document;
+          if (isHello(doc)) {
+            request.reply({ errmsg: 'network error' });
+          } else if (doc.find) {
+            request.reply({
+              cursor: {
+                id: Long.fromNumber(1),
+                ns: 'test.test',
+                firstBatch: [{ _id: 1, name: 'test' }]
+              },
+              ok: 1
+            });
+          }
+        });
+      });
+
+      it('does not set the session on the cursor', function (done) {
+        const topology = new Topology(test.server.hostAddress(), {
+          serverSelectionTimeoutMS: 1000
+        });
+        const cursor = new FindCursor(topology, MongoDBNamespace.fromString('test.test'), {}, {});
+        topology.connect(function () {
+          cursor.next(function () {
+            expect(cursor.session).to.not.exist;
+            topology.close(done);
+          });
+        });
+      });
+    });
+
+    context('when a data bearing server becomes available', function () {
+      beforeEach(function () {
+        let helloCalls = 0;
+        test.server.setMessageHandler(request => {
+          const doc = request.document;
+          if (isHello(doc)) {
+            request.reply(helloCalls > 0 ? { errmsg: 'network error' } : mock.HELLO);
```

Comment:
if I'm reading this correctly, the server is responding with the right response on the very first request, and then with an error on subsequent ones... how does that simulate a data bearing server becoming available? I thought we'd expect to first see an error and then a success for that case. It could be nice to have something here asserting that the expected requests are being made in the expected order or at least some indication of what the expected requests are for each scenario